[prefer-string-starts-ends-with] False positives when slicing

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

<!--
Include a ***minimal*** reproduction case.
The more irrelevant code/config you give, the harder it is for us to investigate.

Please consider creating an isolated reproduction repo to make it easy for the volunteer maintainers debug your issue.
-->

```JSON
{
  "rules": {
    "@typescript-eslint/prefer-string-starts-ends-with": ["error"]
  }
}
```

```TS
const filename = 'carpet.txt';
if (filename.slice(0, -4) === 'car') {
	console.log('file is car.txt');
}
```

<!--
Also include your tsconfig, if you're using type-aware linting
-->

**Expected Result**

No error

**Actual Result**

```
  2:5  error  Use 'String#startsWith' method instead  @typescript-eslint/prefer-string-starts-ends-with
```

**Additional Info**

<!--
Did eslint throw an exception?

Please run your lint again with the --debug flag, and dump the output below.
i.e. eslint --ext ".ts,.js" src --debug
-->

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.4.1` |
| `@typescript-eslint/parser`        | `4.4.1` |
| `TypeScript`                       | `4.0.3` |
| `ESLint`                           | `7.11.0` |
| `node`                             | `14.14.0` |

--- a/packages/eslint-plugin/src/rules/prefer-string-starts-ends-with.ts
+++ b/packages/eslint-plugin/src/rules/prefer-string-starts-ends-with.ts
@@ -575,7 +575,8 @@ export default createRule({
         const isStartsWith =
           !isEndsWith &&
           callNode.arguments.length === 2 &&
-          isNumber(callNode.arguments[0], 0);
+          isNumber(callNode.arguments[0], 0) &&
+          !isNegativeIndexExpression(callNode.arguments[1], node.object);
         if (!isStartsWith && !isEndsWith) {
           return;
         }

--- a/packages/eslint-plugin/tests/rules/prefer-string-starts-ends-with.test.ts
+++ b/packages/eslint-plugin/tests/rules/prefer-string-starts-ends-with.test.ts
@@ -230,6 +230,11 @@ ruleTester.run('prefer-string-starts-ends-with', rule, {
         x.test(s)
       }
     `,
+    `
+      function f(s: string) {
+        s.slice(0, -4) === "car"
+      }
+    `,
   ]),
   invalid: addOptional([
     // String indexing.

--- a/packages/eslint-plugin/tests/rules/prefer-string-starts-ends-with.test.ts
+++ b/packages/eslint-plugin/tests/rules/prefer-string-starts-ends-with.test.ts
@@ -235,6 +235,11 @@ ruleTester.run('prefer-string-starts-ends-with', rule, {
         s.slice(0, -4) === "car"
       }
     `,
+    `
+      function f(x: string, s: string) {
+        x.endsWith('foo') && x.slice(0, -4) === 'bar'
+      }
+    `,
   ]),
   invalid: addOptional([
     // String indexing.

